rs-matter/src/secure_channel/case.rs

```diff
@@ -184,6 +184,16 @@ impl Case {
 
                     session.complete();
 
+                    // Dropping a completed `ReservedSession` will remove the `reserved` flag from the enclosed
+                    // `Session` instance and will thus immediately make the `Session` instance available for use
+                    // by the system.
+                    //
+                    // We need to do this _before_ we send the response to the peer, or else we risk missing
+                    // (dropping) the first messages the peer would send us on the newly-established session,
+                    // as it might start using it right after it receives the response, while it is still marked
+                    // as reserved.
+                    drop(session);
+
```

Comment:
This looks fine to me but one question:

What if we move the handling of the `if self.complete` block in the `Drop` handler for ReservedSesssion in the `complete()` method itself then?

Reply:
We can't really, because of the way Rust destructors work (somebody needs to still tell the destructor not to remove the reserved session from the session manager when `ReservedSession` is dropped, ergo, we still need the `ReservedSession::completed.


**BUT**: We can still do what you suggest, by simply changing `complete(&mut self)` to `complete(self)`:
- This way `ReservedSession::complete` can only be called once
- ...  and immediately after this call the session would be ready, because the destructor of `ReservedSession` would've been called then.

I just did that change (and altered slightly the comments).